File: app/src/core/session_manager.py
"""Session管理器"""
import time
import requests
from typing import Optional

from .njupt_sso import NjuptSso, NjuptSsoException
from .smartclass_client import SmartclassClient
from ..utils.config_manager import config_manager
from ..utils.logger import get_logger

logger = get_logger('session')


class SessionManager:
    """全局登录状态管理"""
    
    def __init__(self):
        self.global_session: Optional[requests.Session] = None
        self.smart_class_client: Optional[SmartclassClient] = None
    
    def set_session(self, session: requests.Session, client: SmartclassClient):
        """保存登录会话"""
        self.global_session = session
        self.smart_class_client = client
    
    def get_session(self) -> Optional[requests.Session]:
        """获取当前会话"""
        return self.global_session
    
    def get_client(self) -> Optional[SmartclassClient]:
        """获取客户端"""
        if self.smart_class_client and not self.is_session_valid():
            logger.info("检测到 Session 过期，尝试自动重连...")
            self.perform_auto_login()
        return self.smart_class_client
    
    def is_session_valid(self) -> bool:
        """检查会话是否有效"""
        if not self.global_session:
            return False
        try:
            r = self.global_session.get("https://njupt.smartclass.cn/", 
                                       allow_redirects=False, timeout=5)
            if r.status_code == 302 and "Login" in r.headers.get("Location", ""):
                return False
            return True
        except:
            return False
    
    def perform_auto_login(self):
        """使用保存的凭证自动登录"""
        config = config_manager.get()
        auth = config_manager.get_auth()
        
        username = ""
        password = ""
        
        if auth and "username" in auth:
            username = auth["username"]
            password = auth.get("password", "")
        elif hasattr(config, "username"):
            username = config.username
            password = getattr(config, "password", "")
        
        if username and password:
            logger.info(f"正在尝试使用保存的账号 {username} 自动登录...")
            try:
                session = requests.Session()
                session.headers.update({
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
                    "Referer": "https://njupt.smartclass.cn/",
                    "Origin": "https://njupt.smartclass.cn",
                    "X-Requested-With": "XMLHttpRequest"
                })
                
                sso = NjuptSso(session)
                sso.login(username, password)
                sso.grant_service("https://njupt.smartclass.cn/Login/SSO")
                session.get("https://njupt.smartclass.cn/", timeout=10)
                
                self.global_session = session
                self.smart_class_client = SmartclassClient(session)
                try:
                    self.smart_class_client.get_csrk_token()
                except:
                    pass
                    
                logger.info(">>> 自动登录成功 (SSO) <<<")
                return True, "登录成功"
            except Exception as e:
                error_msg = str(e)
                logger.error(f"自动登录失败: {e}")
                self.smart_class_client = None
                return False, error_msg
        
        return False, "无保存的凭证"

File: app/internal/smartclass/session_manager.go
package smartclass

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"smartclassdownloader/internal/config"
)

type SessionManager struct {
	mu               sync.RWMutex
	loginMu          sync.Mutex
	config           *config.Service
	newSessionClient func(time.Duration) (*http.Client, error)
	httpClient       *http.Client
	client           *Client
}

func NewSessionManager(service *config.Service) *SessionManager {
	return &SessionManager{config: service, newSessionClient: NewSessionHTTPClient}
}

func (manager *SessionManager) Login(ctx context.Context, username, password string) error {
	manager.loginMu.Lock()
	defer manager.loginMu.Unlock()
	return manager.loginLocked(ctx, username, password)
}

func (manager *SessionManager) loginLocked(ctx context.Context, username, password string) error {
	timeout := time.Duration(manager.config.Current().NetworkTimeoutSeconds) * time.Second
	factory := manager.newSessionClient
	if factory == nil {
		factory = NewSessionHTTPClient
	}
	httpClient, err := factory(timeout)
	if err != nil {
		return err
	}
	sso := NewSSOClient(httpClient)
	if err := sso.Login(ctx, username, password); err != nil {
		return err
	}
	if err := sso.GrantService(ctx, smartclassBaseURL+"/Login/SSO"); err != nil {
		return err
	}
	request, _ := http.NewRequestWithContext(ctx, http.MethodGet, smartclassBaseURL+"/", nil)
	ApplyDefaultHeaders(request)
	if response, err := httpClient.Do(request); err == nil {
		_ = response.Body.Close()
	}
	client := NewClient(httpClient, manager.config)
	if _, err := client.CSRKToken(ctx); err != nil {
		return err
	}
	manager.mu.Lock()
	manager.httpClient = httpClient
	manager.client = client
	manager.mu.Unlock()
	return nil
}

func (manager *SessionManager) AutoLogin(ctx context.Context) error {
	manager.loginMu.Lock()
	defer manager.loginMu.Unlock()
	return manager.autoLoginLocked(ctx)
}

func (manager *SessionManager) autoLoginLocked(ctx context.Context) error {
	username, password, ok, err := manager.config.Credentials(ctx)
	if err != nil {
		return err
	}
	if !ok {
		return fmt.Errorf("无保存的凭证")
	}
	return manager.loginLocked(ctx, username, password)
}

func (manager *SessionManager) Client(ctx context.Context) (*Client, error) {
	client, httpClient := manager.currentPair()
	if client != nil && manager.isValid(ctx, httpClient) {
		return client, nil
	}
	if !manager.config.Current().AutoLogin {
		return nil, fmt.Errorf("登录已失效")
	}

	manager.loginMu.Lock()
	defer manager.loginMu.Unlock()

	client, httpClient = manager.currentPair()
	if client != nil && manager.isValid(ctx, httpClient) {
		return client, nil
	}
	if err := manager.autoLoginLocked(ctx); err == nil {
		client, _ = manager.currentPair()
		return client, nil
	}
	return nil, fmt.Errorf("登录已失效")
}

// ClientForOperation validates the active session before a SmartClass operation.
func (manager *SessionManager) ClientForOperation(ctx context.Context) (*Client, error) {
	return manager.Client(ctx)
}

func (manager *SessionManager) HTTPClient(ctx context.Context) (*http.Client, error) {
	_, httpClient, err := manager.ClientPair(ctx)
	if err != nil {
		return nil, err
	}
	return httpClient, nil
}

func (manager *SessionManager) ClientPair(ctx context.Context) (*Client, *http.Client, error) {
	if _, err := manager.Client(ctx); err != nil {
		return nil, nil, err
	}
	manager.mu.RLock()
	defer manager.mu.RUnlock()
	if manager.client == nil || manager.httpClient == nil {
		return nil, nil, fmt.Errorf("登录已失效")
	}
	return manager.client, manager.httpClient, nil
}

func (manager *SessionManager) IsValid(ctx context.Context) bool {
	_, current := manager.currentPair()
	return manager.isValid(ctx, current)
}

func (manager *SessionManager) isValid(ctx context.Context, current *http.Client) bool {
	if current == nil {
		return false
	}
	probe := *current
	probe.CheckRedirect = func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }
	request, _ := http.NewRequestWithContext(ctx, http.MethodGet, smartclassBaseURL+"/", nil)
	ApplyDefaultHeaders(request)
	response, err := probe.Do(request)
	if err != nil {
		return false
	}
	defer response.Body.Close()
	return response.StatusCode != http.StatusFound || !strings.Contains(strings.ToLower(response.Header.Get("Location")), "login")
}

func (manager *SessionManager) Clear() {
	manager.mu.Lock()
	manager.client = nil
	manager.httpClient = nil
	manager.mu.Unlock()
}

func (manager *SessionManager) currentPair() (*Client, *http.Client) {
	manager.mu.RLock()
	defer manager.mu.RUnlock()
	return manager.client, manager.httpClient
}

func HeadersAndCookies(client *http.Client, target string) []string {
	headers := []string{
		"User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
		"Referer: " + smartclassBaseURL + "/",
	}
	parsed, err := url.Parse(target)
	if err != nil || client == nil || client.Jar == nil {
		return headers
	}
	cookies := client.Jar.Cookies(parsed)
	parts := make([]string, 0, len(cookies))
	for _, cookie := range cookies {
		parts = append(parts, cookie.Name+"="+cookie.Value)
	}
	if len(parts) > 0 {
		headers = append(headers, "Cookie: "+strings.Join(parts, "; "))
	}
	return headers
}
